### ISSUE TITLE
Wildcards in redirect URI do not work as expected

### ISSUE THREAD

**Issue by tschumwalleye** [workos/authkit-nextjs#24]
I have branch URLs from Vercel preview deployments. 

My `WORKOS_REDIRECT_URI` is configured as a wildcard in my dashboard following the guidelines here: [WorkOS Docs](https://workos.com/docs/sso/redirect-uris/wildcard-characters).

After migrating to authkit-nextjs, redirected users land on a url where the `*` is replaced with `%2A`. 

https://*.example.com -> https://%2A.example.com

I would be able to pass in the URL at run-time if that was an option, which is how I handled this behavior before migrating. I would take advantage of the `VERCEL_URL` environment variable when calling the `getAuthorizationUrl` method.

**Comment by PaulAsjes** [comment]
Thanks for reporting this, that's definitely unintentional. We'll look into this and get back to you!

**Comment by benoitgrelard** [comment]
@tschumwalleye I'd like to ask a few questions to make sure I understand the issue:
- You have added `https://*.example.com` as a redirect URI in the dashboard?
-  What is the value of `WORKOS_REDIRECT_URI` that you pass in? Is it the same URL?

I feel like the issue is that you are unable to pass in the actual URL (one that would match the wildcard scheme) at runtime. Is that correct?

If so, I think we just need to add an option to the `getSignInUrl` so it can be passed at runtime.

**Comment by benoitgrelard** [comment]
I think `signOut` is different though, the URL it redirects to is configured in the dashboard.

<img width="606" alt="CleanShot 2024-04-05 at 14 18 15@2x" src="https://github.com/workos/authkit-nextjs/assets/1539897/f1941505-62d8-45c8-9fba-5a6288eb552f">

**Comment by tschumwalleye** [comment]
That is correct, this leaves me with the same issue of being unable to pass in an accurate URL for preview deployments. I can bypass using the signOut function by calling `cookies().delete("wos-session")` myself, then redirecting. My concerns with this are:
- If that string changes in the future
- Are idp initiated SSO users redirected to a different logout page? (Which would be provided by `getLogoutUrl`)

**Comment by jxdp** [comment]
[vercel-system-vars]: https://vercel.com/docs/projects/environment-variables/system-environment-variables#framework-environment-variables

@tschumwalleye I found a way to work around this using the [experimental instrumentation hook](https://nextjs.org/docs/app/building-your-application/optimizing/instrumentation).

It isn't elegant, but it works...

```ts
// instrumentation.ts
export function register() {
  if (process.env.WORKOS_REDIRECT_URI && process.env.VERCEL_PROJECT_PRODUCTION_URL) {
    const pathname = new URL(process.env.WORKOS_REDIRECT_URI).pathname;
    process.env.WORKOS_REDIRECT_URI = `https://${process.env.VERCEL_PROJECT_PRODUCTION_URL}${pathname}`;
  }
}
```

What I would like to see is a new export that the NextAuth pattern:

```tsx
import NextAuth from "next-auth"
 
export const { handlers, signIn, signOut, auth } = NextAuth({
  // ...
})
```

which would look along the lines of the following. This could be exposed under a new export without impacting the existing API, or as a breaking change to the existing API.

```tsx
// authkit.ts

import { AuthKit } from "@workos-inc/authkit-nextjs"; 
// or import { AuthKit } from "@workos-inc/authkit-nextjs/some-new-export";
 
export const {
  authkitMiddleware,
  getSignInUrl,
  getSignOutUrl,
  getUser,
  handleAuth,
  signOut,
  Impersonate,
} = AuthKit({
  redirectUrl: process.VERCEL_PROJECT_PRODUCTION_URL ?
    `https://${process.VERCEL_PROJECT_PRODUCTION_URL}/api/auth` :
     "http://localhost:3000/api/auth",
});
```

Which provides essentially the same level of ease, for example:

`import { signOut } from "@workos-inc/authkit-nextjs"` becomes `import { signOut } from "@/path/to/authkit"`

**Comment by jxdp** [comment]
@PaulAsjes I have exactly the same issue.

Have tried many workarounds:
- setting env variables at runtime
- applying a custom webpack plugin
- using `.env` files with  `WORKOS_REDIRECT_URI="https://${VERCEL_URL}/my/redirect/path"`)

But none of these work due to how Vercel/Next handles environment variables on the edge/middleware runtime.

I think what I'd like to see here is a solution that allows for all the config to be passed in at runtime to create an object, from which everything else can be imported (analogous to how NextAuth works), e.g.

```ts
export const { authkitMiddleware, signOut, getUser, ... } = AuthKit({
  redirectUri: ...,
  cookieSecret: ...,
  clientId: ...,
  clientSecret: ...,
});
```

**Comment by PaulAsjes** [comment]
Fixed in v0.11.0.

**Comment by moroshko** [comment]
@PaulAsjes Does `getSignInUrl()` take into account the `redirectUri` passed to `authkitMiddleware`?

I'm using `@workos-inc/authkit-nextjs@1.3.0`, added `https://phonic-*-phonic.vercel.app/api/auth/callback` as a Redirect URI in `Redirects -> Sign-in callback`, and passing the `redirectUri` to `authkitMiddleware`:

```ts
export function middleware(request: NextRequest, event: NextFetchEvent) {
  console.log({ redirectUri: `${siteUrl}/api/auth/callback` }); // https://phonic-git-sticky-mic-phonic.vercel.app/api/auth/callback

  return authkitMiddleware({
    redirectUri: `${siteUrl}/api/auth/callback`,
  })(request, event);
}
```

However, the URL that `getSignInUrl()` returns doesn't seem to have a `redirect_uri`:

```
https://api.workos.com/user_management/authorize?client_id=...&provider=authkit&response_type=code&screen_hint=sign-in
```

What am I missing here?

**Comment by nicknisi** [comment]
Hey @moroshko, this was fixed in #209 and released in [v2.0.1](https://github.com/workos/authkit-nextjs/releases/tag/v2.0.1). Thanks!

**Comment by moroshko** [comment]
Thanks @nicknisi, I should have checked the latest version 🤦‍♂️
I can confirm it works as expected with v2.0.1.
